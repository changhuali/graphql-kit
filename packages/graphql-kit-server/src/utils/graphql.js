import { getNamedType, isEnumType, isScalarType, isUnionType } from "graphql"

/**
 * extract field info from field map
 * @param {import("graphql").GraphQLFieldMap<any, any> | import("graphql").GraphQLInputFieldMap} fieldMap - input definition node
 * @returns {any}
 */
export function extractFieldInfoFromFieldMap(fieldMap) {
  /** @type {Record<string, any>} */
  const result = {}
  Object.keys(fieldMap).forEach((key) => {
    const field = fieldMap[key]
    const namedType = getNamedType(field.type)
    result[key] = {
      description: field.description || '',
      type: field.type.toString(),
      directives: field.astNode?.directives,
      ...getTypeDefFromNamedType(namedType)
    }
  })
  return result
}

/**
 * extract field info from named type
 * @param {import("graphql").GraphQLNamedType} namedType - named type
 */
export function getTypeDefFromNamedType(namedType) {
  /** @type {Record<string, any>} */
  const result = {}
  if (isEnumType(namedType)) {
    result.typeDef = namedType.getValues().map(item => ({
      name: item.name,
      description: item.description || '',
      directives: item.astNode?.directives || [],
      value: item.value,
    }))
  } else if (isUnionType(namedType)) {
    result.typeDef = namedType.getTypes().map(item => {
      return extractFieldInfoFromFieldMap(item.getFields())
    }).reduce((result, item) => {
      return {...result, ...item}
    }, {})
  } else if (isScalarType(namedType)) {
    // Do nothing
  } else {
    result.typeDef = extractFieldInfoFromFieldMap(namedType.getFields())
  }
  return result
}

/**
 * extract operation info from graphql field
 * @param {import('graphql').GraphQLField<any, any>} graphQLField 
 * @param {import('graphql').GraphQLSchema} rawSchema 
 * @returns {object}
 * TODO:avoid circular ref
 */
 export function extractOperationInfoFromGraphQLField(graphQLField, rawSchema) {
  const typeMap = rawSchema.getTypeMap()
  const returnTypeValue = typeMap[getNamedType(graphQLField.type).name]
  return {
    name: graphQLField?.name,
    description: graphQLField?.description || '',
    directives: graphQLField.astNode?.directives || [],
    arguments: graphQLField?.args.map(item => {
      const namedType = getNamedType(item.type)
      const namedTypeValue = typeMap[namedType.name]
      return {
        name: item.name,
        description: item.description || '',
        directives: item.astNode?.directives || [],
        type: item.type.toString(),
        ...getTypeDefFromNamedType(namedTypeValue),
      }
    }),
    return: {
      name: graphQLField.name,
      directives: returnTypeValue.astNode?.directives,
      description: returnTypeValue.astNode?.description?.value || '',
      type: graphQLField?.type,
      ...getTypeDefFromNamedType(returnTypeValue),
    },
  }
}
