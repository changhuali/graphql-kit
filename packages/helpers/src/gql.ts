import type { OperationArgument, TypedOperation, TypeDef } from './interface'

/**
 * generate a space string
 * @param count - the str length
 */
export const genSpace = (count: number) => {
  return ' '.repeat(count)
}

/**
 * generate an outer arguments string
 * @param args - the arguments of operation
 */
const _genOuterArgumentsStr = (args: OperationArgument[]) => {
  if (args.length) {
    return `(${args.map(item => `$${item.name}: ${item.type}`).join(', ')})`
  }
  return ''
}

/**
 * generate an inner arguments string
 * @param args - operation's arguments
 */
const _genInnerArgumentsStr = (args: OperationArgument[]) => {
  if (args.length) {
    return `(${args.map(item => `${item.name}: $${item.name}`).join(', ')})`
  }
  return ''
}

/**
 * generate a typeDef gql
 * @param typeDef - the type of operation
 * @param indent - the indent of gql
 */
const _genTypeDefStr = (typeDef: TypeDef, indent: number) => {
  if (!typeDef) return '\n'
  let str = ''
  if (Array.isArray(typeDef)) {
    str += '\n'
  } else {
    str += ` {\n`
    const fields = Object.entries(typeDef)
    fields.forEach(([field, fieldInfo]) => {
      str += `${genSpace(indent)}${field}`
      str += _genTypeDefStr(fieldInfo.typeDef, indent + 2)
    })
    str += `${genSpace(indent - 2)}}\n`
  }
  return str
}

/**
 * generate a gql string
 * @param operation - operation which need to be converted into a gql
 * @param indent - the indent of gql
 */
export const genGQLStr = (operation: TypedOperation, indent = 2) => {
  const returnTypeDef = operation.return.typeDef
  const descriptionStr = operation.description
    ? `# ${operation.description}\n`
    : ''
  return `${descriptionStr}${operation.type}${genSpace(1)}${
    operation.name
  }${_genOuterArgumentsStr(operation.arguments)}${genSpace(1)}{\n${genSpace(
    indent,
  )}${operation.name}${_genInnerArgumentsStr(operation.arguments)}${
    returnTypeDef ? _genTypeDefStr(returnTypeDef, indent + indent) : '\n'
  }}`
}

/**
 * generate a group of gqls
 * @param groupName - the name of group
 * @param operations - the operations in a group
 */
export const genGQLStrInGroup = (
  groupName: string,
  operations: TypedOperation[],
) => {
  return operations.reduce((str, operation, index) => {
    const isLast = index === operations.length - 1
    return str + genGQLStr(operation) + (isLast ? '\n' : '\n\n')
  }, `# ${groupName} \n\n`)
}
