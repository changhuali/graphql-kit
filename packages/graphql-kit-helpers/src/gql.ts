import { OperationTypeNode } from 'graphql'
import {
  Operation,
  OperationArgument,
  TypedOperation,
  ScalarMap,
  TypeDef,
} from './types'

/**
 * generate a space string
 * @param count the str length
 */
export const genSpace = (count: number) => {
  return ' '.repeat(count)
}

/**
 * generate an outer arguments string
 * @param args the arguments of operation
 */
const _genOuterArgumentsStr = (args: OperationArgument[]) => {
  if (args.length) {
    return `(${args.map(item => `$${item.name}: ${item.type}`).join(', ')})`
  }
  return ''
}

/**
 * generate an inner arguments string
 * @param args operation's arguments
 */
const _genInnerArgumentsStr = (args: OperationArgument[]) => {
  if (args.length) {
    return `(${args.map(item => `${item.name}: $${item.name}`).join(', ')})`
  }
  return ''
}

/**
 * generate a typeDef gql
 * @param args the type of operation
 * @param indent the indent of gql
 */
const _genTypeDefStr = (typeDef: TypeDef, indent: number) => {
  if (!typeDef) return '\n'
  let str = ''
  if (Array.isArray(typeDef)) {
    str += '\n'
  } else {
    str += ` {\n`
    const fields = Object.entries(typeDef)
    fields.forEach(([field, fieldInfo]) => {
      str += `${genSpace(indent)}${field}`
      str += _genTypeDefStr(fieldInfo.typeDef, indent + 2)
    })
    str += `${genSpace(indent - 2)}}\n`
  }
  return str
}

/**
 * generate a gql string
 * @param operationType the type of operation
 * @param operation operation which need to be converted into a gql
 * @param indent the indent of gql
 */
export const genGQLStr = (
  operationType: OperationTypeNode,
  operation: Operation,
  indent = 2,
) => {
  const returnTypeDef = operation.return.typeDef
  const descriptionStr = operation.description
    ? `# ${operation.description}\n`
    : ''
  return `${descriptionStr}${operationType}${genSpace(1)}${
    operation.name
  }${_genOuterArgumentsStr(operation.arguments)}${genSpace(1)}{\n${genSpace(
    indent,
  )}${operation.name}${_genInnerArgumentsStr(operation.arguments)}${
    returnTypeDef ? _genTypeDefStr(returnTypeDef, indent + indent) : '\n'
  }}`
}

/**
 * generate a variables object
 * @param args the arguments of operation
 * @param scalarMap a map contains the default value of scalar type
 */
export const genVariables = (
  args: OperationArgument[],
  scalarMap: ScalarMap,
) => {
  const variables: Record<string, unknown> = {}
  args.forEach(({ name, type, typeDef }) => {
    let defaultValue
    if (!typeDef) {
      // scalar type
      defaultValue = scalarMap[type]
    } else if (Array.isArray(typeDef)) {
      // enum type
      defaultValue = typeDef[0].value
    } else {
      // object type
      const subArgs = Object.entries(typeDef).map(
        ([fieldName, fieldTypeDef]) => {
          return {
            name: fieldName,
            ...fieldTypeDef,
          }
        },
      )
      defaultValue = genVariables(subArgs, scalarMap)
    }
    variables[name] = defaultValue
  })
  return variables
}

/**
 * generate a group of gqls
 * @param groupName the name of group
 * @param operations the operations in a group
 */
export const genGQLStrInGroup = (
  groupName: string,
  operations: TypedOperation[],
) => {
  return operations.reduce((str, { type, ...operation }, index) => {
    const isLast = index === operations.length - 1
    return str + genGQLStr(type, operation) + (isLast ? '\n' : '\n\n')
  }, `# ${groupName} \n\n`)
}
