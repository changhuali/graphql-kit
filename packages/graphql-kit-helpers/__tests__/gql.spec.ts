import { readFileSync } from 'fs'
import { URL } from 'url'
import { buildSchema, OperationTypeNode } from 'graphql'
import { genSpace, genGQLStr, genGQLStrInGroup, genVariables } from '@src/gql'
import {
  getOperationFromGraphQLField,
  getOperationsBySchema,
  groupOperations,
} from '@src/operation'

/********** genSpace start **********/
describe('genSpace', () => {
  it('a string consisting of a given number of spaces should be generated', () => {
    expect(genSpace(0)).toBe('')
    expect(genSpace(10)).toBe(' '.repeat(10))
  })
})
/********** genSpace end ************/

/********** genGQLStr start **********/
const schemaPath = new URL('./schema.graphql', import.meta.url).pathname
const schemaString = readFileSync(schemaPath, {
  encoding: 'utf-8',
})
const schema = buildSchema(schemaString)
const allOperations = getOperationsBySchema(schema)
const operationsAfterBeingGrouped = groupOperations(allOperations)

describe('genGQLStr', () => {
  const fieldsMap = {
    query: schema.getQueryType()?.getFields() || {},
    mutation: schema.getMutationType()?.getFields() || {},
    subscription: schema.getSubscriptionType()?.getFields() || {},
  }
  Object.entries(fieldsMap).forEach(([operationType, graphQLFields]) => {
    Object.values(graphQLFields).forEach(graphQLField => {
      it(`a gql string for ${graphQLField.name} ${operationType} should be generated`, () => {
        const operation = getOperationFromGraphQLField(graphQLField, schema)
        const gqlString = genGQLStr(OperationTypeNode.QUERY, operation)
        expect(gqlString).toMatchSnapshot()
      })
    })
  })
})
/********** genGQLStr start ************/

/********** genGQLStrInGroup start ************/
describe('genGQLStrInGroup', () => {
  Object.entries(operationsAfterBeingGrouped).forEach(
    ([groupName, operations]) => {
      it(`gql strings for group ${groupName} should be generated`, () => {
        const gqlStrings = genGQLStrInGroup(groupName, operations)
        expect(gqlStrings).toMatchSnapshot()
      })
    },
  )
})
/********** genGQLStrInGroup end **************/

/********** genVariables start ************/
describe('genVariables', () => {
  allOperations.forEach(operation => {
    it(`a variable for the operation ${operation.name} ${operation.type} should be generated`, () => {
      expect(genVariables(operation.arguments, {})).toMatchSnapshot()
    })
  })
})
/********** genVariables end **************/
