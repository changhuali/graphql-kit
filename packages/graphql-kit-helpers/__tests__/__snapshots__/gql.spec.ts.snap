// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genGQLStr a gql string for user mutation should be generated 1`] = `
"# 用户: 修改用户信息
query user($user: UserMutationInput) {
  user(user: $user)
}"
`;

exports[`genGQLStr a gql string for user query should be generated 1`] = `
"# 用户: 根据用户id获取用户信息
query user($userId: ID) {
  user(userId: $userId) {
    id
    name
    age
  }
}"
`;

exports[`genGQLStrInGroup gql strings for group 用户 should be generated 1`] = `
"# 用户 

# 根据用户id获取用户信息
query user($userId: ID) {
  user(userId: $userId) {
    id
    name
    age
  }
}

# 修改用户信息
mutation user($user: UserMutationInput) {
  user(user: $user)
}
"
`;

exports[`genVariables a variable for the operation user mutation should be generated 1`] = `
Object {
  "user": Object {
    "age": undefined,
    "id": undefined,
    "name": undefined,
  },
}
`;

exports[`genVariables a variable for the operation user query should be generated 1`] = `
Object {
  "userId": undefined,
}
`;
